src/main/java/com/saucelabs/ci/sauceconnect/SauceConnectFourManager.java

```diff
@@ -180,7 +180,11 @@ protected Process prepAndCreateProcess(String username, String apiKey, int port,
                 File unzipDirectory = getUnzipDir(workingDirectory, operatingSystem);
                 sauceConnectBinary = new File(unzipDirectory, operatingSystem.getExecutable());
                 if (!sauceConnectBinary.exists()) {
```

Comment:
@valfirst Do we need this `if` statement if it is already in `synchronized` block?

Reply:
yes, it's needed for performance optimization: if SC binary exists,`synchronized` block is not run, and execution is not blocked, but without this `if` the execution will be blocked by a single thread any time, this code is invoked.

also the inner `if` is required to avoid possible case when 2 threads passed the first `if`, one thread executed `synchronized` block and unpacked SC binary (while the second thread was waiting), and when the second thread entered `synchronized` block, SC binary already exists

Reply:
@adambabik the current implementation is very similar to double-checked locking algorithm (it could be considered broken for singleton implementation, but in this case it works perfectly since we deal with files)

Reply:
Thanks for explanation. It makes sense in this case.